### PR TITLE
user defined transports

### PR DESCRIPTION
**Pull request by amitmurthy**
This builds on #9309 and supports user definable transports.
It supersedes #9046 - the discussion there outlines much of the motivations for this PR.

To implement this, custom cluster managers would need to provide their own 
`connect(manager::FooManager, pid::Integer, config::WorkerConfig)` method

`connect` should return a pair of `AsyncStream` objects, one for reading data sent from worker `pid`, and the other to write data that needs to be sent to worker `pid`. custom cluster managers can use an in-memory `BufferStream` as the plumbing to ferry data between the custom, non AsyncStream transport and the Julia parallel infrastructure. 

A `BufferStream` wraps a `PipeBuffer` and condition variables to make a waitable stream.  

`examples/clustermanager/0mq` is an example of how they are used to setup a star network with a 0MQ broker in the middle.

`connect` is optional, and the default implementation is based on TCP as a transport mechanism.

Another optional method is `kill(manager::FooManager, pid::Int, config::WorkerConfig)`, which is called to remove a worker from the cluster.

Two example implementations are provided:
- `examples/clustermanager/simple` shows the use of unix domain sockets as transport
- `examples/clustermanager/0mq` shows the use of 0MQ as transport

One thought is to move the examples to package `ClusterManagers.jl` instead of adding it here. Have kept it in this PR so that folks can have a look. Will move it based on feedback.

Documentation still needs to updated.

### CODE REVIEW

**Comment by amitmurthy** [comment]
Documentation has been added. In the absence of any concerns I'll merge this over the next day or two.

cc: @JeffBezanson

**Comment by tkelman** [comment]
Do the examples here get run by `test/examples.jl`?

**Comment by ViralBShah** [comment]
Wouldn't that introduce new dependencies on Base, such as zeromq? Perhaps the test can be conditional on finding zeromq.

**Comment by tkelman** [comment]
Good point. Would the simple one be quick and painless to test, or does it require some infrastructure?

**Comment by amitmurthy** [comment]
Good idea, I'll add the simple example (unix domain sockets) to test. And maybe the 0mq one with a dummy module at least to ensure that the code loads.

**Comment by tkelman** [comment]
should probably be `@unix_only`

**Comment by tkelman** [comment]
And would that test be possible in environments with limited/no internet access like distribution buildbots? As long it's not more demanding in resources than our current socket or parallel tests it should be fine.

**Comment by amitmurthy** [comment]
As long as networking is enabled, we can test by starting just 2 additional workers - so that would be a total of 3 julia processes.

**Comment by amitmurthy** [comment]
@tkelman - could you take a look now? I had to add a line to the Makefile to copy these examples to the build doc dir. The simple manager is run `@unix_only` out of process since mixing multiple cluster managers in the same setup is not fully supported (workers started via different cluster managers do not know how to connect to each other). As for 0mq, only code loading is tested.

**Comment by tkelman** [comment]
LGTM. Not sure whether this could cause any issues on distribution buildbots, but there's an easy way to find out.

I might've done the temporary-file test using a `with`-`do` block, but we're not too rigorous about cleaning up after all the other tests if there happens to be a failure.

File: test/examples.jl [hunk 1]
```diff
@@ -35,9 +35,47 @@ include(joinpath(dir, "queens.jl"))
...
 
 include(joinpath(dir, "preduce.jl"))
 
 include(joinpath(dir, "wordcount.jl"))
+
+# the 0mq clustermanager depends on package ZMQ. Just making sure the
+# code loads using a stub module definition for ZMQ.
+zmq_found=true
+try
+    using zmq
```
Comment (tkelman) [inline test/examples.jl]: wait, this should probably be capitalized

Comment (amitmurthy) [inline test/examples.jl]: Yes, good catch.

**Pull request merged**